cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1620 @@
+
+                if (pool.total_size == pairs[1])
+                    continue;
+
+                pool.total_size = pairs[1];
+
+                if (pool.initialized)
+                {
+                    updateBufferPoolToDb(poolName, pool);
+                }
+                else
+                {
+                    updateBufferPoolToDb(poolName, pool);
+                    pool.initialized = true;
+                }
+
+                SWSS_LOG_NOTICE("Buffer pool %s had been updated with new size [%s]", poolName.c_str(), pool.total_size.c_str());
+            }
+            else
+            {
+                SWSS_LOG_INFO("Buffer pool debug info %s", i.c_str());
+            }
+        }
+    }
+    catch (...)
+    {
+        SWSS_LOG_WARN("Lua scripts for buffer calculation were not executed successfully");
+    }
+}
+
+void BufferMgrDynamic::checkSharedBufferPoolSize()
+{
+    // PortInitDone indicates all steps of port initialization has been done
+    // Only after that does the buffer pool size update starts
+    if (!m_portInitDone)
+    {
+        vector<FieldValueTuple> values;
+        if (m_applPortTable.get("PortInitDone", values))
+        {
+            SWSS_LOG_NOTICE("Buffer pools start to be updated");
+            m_portInitDone = true;
+        }
+        else
+        {
+            if (m_firstTimeCalculateBufferPool)
+            {
+                // It's something like a placeholder especially for warm reboot flow
+                // without all buffer pools created, buffer profiles are unable to be cureated,
+                // which in turn causes buffer pgs and buffer queues unable to be created,
+                // which prevents the port from being ready and eventually fails the warm reboot
+                // After the buffer pools are created for the first time, we won't touch it
+                // until portInitDone
+                // Eventually, the correct values will pushed to APPL_DB and then ASIC_DB
+                recalculateSharedBufferPool();
+                m_firstTimeCalculateBufferPool = false;
+                SWSS_LOG_NOTICE("Buffer pool update defered because port is still under initialization, start polling timer");
+            }
+
+            return;
+        }
+    }
+
+    if (!m_mmuSize.empty())
+        recalculateSharedBufferPool();
+}
+
+// For buffer pool, only size can be updated on-the-fly
+void BufferMgrDynamic::updateBufferPoolToDb(const string &name, const buffer_pool_t &pool)
+{
+    vector<FieldValueTuple> fvVector;
+
+    if (pool.ingress)
+        fvVector.emplace_back(make_pair("type", "ingress"));
+    else
+        fvVector.emplace_back(make_pair("type", "egress"));
+
+    fvVector.emplace_back(make_pair("mode", pool.mode));
+
+    SWSS_LOG_INFO("Buffer pool %s is initialized", name.c_str());
+
+    fvVector.emplace_back(make_pair("size", pool.total_size));
+
+    m_applBufferPoolTable.set(name, fvVector);
+
```

Comment:
Can we do error check for the max buffer pool sizes for all the configured pools and that shouldn't exceed max mmu size?

Reply:
It’s a bit difficult to achieve that because all asic vendors don’t share the same logic of testing the thresholds.
I suggest doing this in these ways:
- pools with static size: make sure the configuration is correct
- pools dynamically calculated: check that in the lua plug in

Reply:
Checking added